File: src/languages/python/assets/tests/testcases/socket_test.py
from __future__ import annotations

import unittest

from ffi.socket import MessageSocket
from stream import SocketStream


class MessageSocketTests(unittest.TestCase):
    def test_message_socket_roundtrip(self) -> None:
        left, right = SocketStream.pair()
        try:
            sender = MessageSocket(left)
            receiver = MessageSocket(right)

            sender.send_text("hello")
            self.assertEqual(receiver.receive_text(), "hello")
        finally:
            left.close()
            right.close()

    def test_message_socket_handles_heavy_fragmentation(self) -> None:
        left, right = SocketStream.pair()
        right.max_read_size = 1
        try:
            sender = MessageSocket(left)
            receiver = MessageSocket(right)

            sender.send_text("fragmented message")
            self.assertEqual(receiver.receive_text(), "fragmented message")
        finally:
            left.close()
            right.close()

    def test_message_socket_parses_multiple_messages_from_one_chunk(self) -> None:
        left, right = SocketStream.pair()
        try:
            receiver = MessageSocket(right)
            left.write(b"5:hello,5:world,")

            self.assertEqual(receiver.receive_text(), "hello")
            self.assertEqual(receiver.receive_text(), "world")
        finally:
            left.close()
            right.close()

    def test_message_socket_eof_returns_none(self) -> None:
        left, right = SocketStream.pair()
        try:
            receiver = MessageSocket(right)
            left.close()
            self.assertIsNone(receiver.receive_text())
        finally:
            right.close()

    def test_message_socket_rejects_non_digit_length(self) -> None:
        left, right = SocketStream.pair()
        try:
            receiver = MessageSocket(right)
            left.write(b"x:abc,")

            with self.assertRaisesRegex(ValueError, "Invalid netstring"):
                receiver.receive()
        finally:
            left.close()
            right.close()

    def test_message_socket_rejects_missing_comma(self) -> None:
        left, right = SocketStream.pair()
        try:
            receiver = MessageSocket(right)
            left.write(b"3:abc.")

            with self.assertRaisesRegex(ValueError, "missing comma"):
                receiver.receive()
        finally:
            left.close()
            right.close()


if __name__ == "__main__":
    unittest.main()

File: src/languages/cpp/assets/tests/testcases/socket_test.cpp
#include <cassert>
#include <string>

#include "stream.hpp"

int main() {
    using ptip_ffi::MessageSocket;

    {
        auto [left, right] = ptip_ffi::tests::MemoryDuplexStream::pair();
        MessageSocket sender(left);
        MessageSocket receiver(right);

        sender.send("hello");
        assert(receiver.receive() == "hello");
    }

    {
        auto [left, right] = ptip_ffi::tests::MemoryDuplexStream::pair();
        right.set_read_limit(1);
        MessageSocket sender(left);
        MessageSocket receiver(right);

        sender.send("fragmented message");
        assert(receiver.receive() == "fragmented message");
    }

    {
        auto [left, right] = ptip_ffi::tests::MemoryDuplexStream::pair();
        left.write("5:hello,5:world,", 16);
        MessageSocket receiver(right);
        assert(receiver.receive() == "hello");
        assert(receiver.receive() == "world");
    }

    {
        auto [left, right] = ptip_ffi::tests::MemoryDuplexStream::pair();
        left.write("x:abc,", 6);
        MessageSocket receiver(right);
        bool threw = false;
        try {
            receiver.receive();
        } catch (const std::exception&) {
            threw = true;
        }
        assert(threw);
    }

    return 0;
}
